File: mirror/tier2-ruby/showcase/pattern_binding.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Pattern Matching
# Module: pattern_binding.rb
# Purpose: Demonstrate pattern binding and destructuring

def destructure_tuple(data)
  if data.is_a?(Array) && data.size == 3
    a, b, c = data
    {first: a, second: b, third: c}
  end
end

def destructure_hash(data)
  if data.is_a?(Hash)
    name = data["name"] || "unknown"
    age = data["age"] || 0
    {person: name, age: age}
  end
end

def destructure_list(data)
  if data.is_a?(Array) && data.size >= 2
    head = data[0]
    tail = data[1..-1]
    {head: head, tail: tail}
  end
end

def bind_and_match(obj)
  if obj.is_a?(Hash)
    x = obj["x"]
    y = obj["y"]
    if x && y
      {bound_x: x, bound_y: y, sum: x + y}
    end
  end
end

result = {
  tuple_destructure: destructure_tuple([1, 2, 3]),
  hash_destructure: destructure_hash({"name" => "Alice", "age" => 30}),
  list_destructure: destructure_list([10, 20, 30, 40]),
  bind_and_match: bind_and_match({"x" => 5, "y" => 10}),
}

puts result.inspect

File: mirror/tier2-dart/showcase/pattern_binding.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Pattern Matching
// Module: pattern_binding.dart

Map<String, dynamic>? destructureTuple(List<dynamic> data) {
  if (data.length == 3) {
    final [a, b, c] = data;
    return {'first': a, 'second': b, 'third': c};
  }
  return null;
}

Map<String, dynamic>? destructureHash(Map<String, dynamic>? data) {
  if (data == null) return null;
  
  final name = data['name'] ?? 'unknown';
  final age = data['age'] ?? 0;
  return {'person': name, 'age': age};
}

Map<String, dynamic>? destructureList(List<dynamic> data) {
  if (data.isNotEmpty) {
    final head = data[0];
    final tail = data.sublist(1);
    return {'head': head, 'tail': tail};
  }
  return null;
}

Map<String, dynamic>? bindAndMatch(Map<String, dynamic>? obj) {
  if (obj != null && obj.containsKey('x') && obj.containsKey('y')) {
    final x = obj['x'] as int;
    final y = obj['y'] as int;
    return {'boundX': x, 'boundY': y, 'sum': x + y};
  }
  return null;
}

void main() {
  final result = <String, dynamic>{
    'tupleDestructure': destructureTuple([1, 2, 3]),
    'hashDestructure': destructureHash({'name': 'Alice', 'age': 30}),
    'listDestructure': destructureList([10, 20, 30, 40]),
    'bindAndMatch': bindAndMatch({'x': 5, 'y': 10}),
  };
  print(result);
}
